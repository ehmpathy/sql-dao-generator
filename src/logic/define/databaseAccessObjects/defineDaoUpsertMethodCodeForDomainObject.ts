import { camelCase } from 'change-case';
import { DomainObjectMetadata } from 'domain-objects-metadata';

import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { castDomainObjectNameToDaoName } from './castDomainObjectNameToDaoName';
import { defineOutputTypeOfFoundDomainObject } from './defineOutputTypeOfFoundDomainObject';
import {
  defineQueryFunctionInputExpressionForDomainObjectProperty,
  GetTypescriptCodeForPropertyContext,
} from './defineQueryFunctionInputExpressionForDomainObjectProperty';
import { defineQueryInputExpressionForSqlSchemaProperty } from './defineQueryInputExpressionForSqlSchemaProperty';

export const defineDaoUpsertMethodCodeForDomainObject = ({
  domainObject,
  sqlSchemaRelationship,
  allSqlSchemaRelationships,
}: {
  domainObject: DomainObjectMetadata;
  sqlSchemaRelationship: SqlSchemaToDomainObjectRelationship;
  allSqlSchemaRelationships: SqlSchemaToDomainObjectRelationship[];
}) => {
  // define some useful constants
  const sqlSchemaName = sqlSchemaRelationship.name.sqlSchema;
  const hasUuidProperty = !!domainObject.properties.uuid;

  // define the imports
  const imports = [
    // always present imports
    `import { HasId${hasUuidProperty ? ', HasUuid' : ''} } from 'simple-type-guards';`,
    "import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';",
    `import { ${domainObject.name} } from '$PATH_TO_DOMAIN_OBJECT';`, // import this domain object; note: higher level function will swap out the import path
    "import { log } from '$PATH_TO_LOG_OBJECT';",
    `import { sqlQueryUpsert${domainObject.name} } from '$PATH_TO_GENERATED_SQL_QUERY_FUNCTIONS';`,
    ...sqlSchemaRelationship.properties
      .filter(
        (propertyRelationship) =>
          propertyRelationship.sqlSchema.reference?.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING, // this property is nested directly
      )
      .map((propertyRelationship) => {
        const nameOfDaoToImport = castDomainObjectNameToDaoName(propertyRelationship.sqlSchema.reference!.of.name);
        return `import { ${nameOfDaoToImport} } from '../${nameOfDaoToImport}';`;
      })
      .sort(),
  ];

  // define the query input expressions
  const queryInputExpressions: string[] = Object.values(sqlSchemaRelationship.properties).map(
    ({ sqlSchema: sqlSchemaProperty, domainObject: domainObjectProperty }) =>
      defineQueryInputExpressionForSqlSchemaProperty({
        sqlSchemaName,
        sqlSchemaProperty,
        domainObjectProperty,
        allSqlSchemaRelationships,
      }),
  );

  // define the queryFunctionInputExpressions
  const queryFunctionInputExpressions: string[] = Object.values(sqlSchemaRelationship.properties).map(
    ({ sqlSchema: sqlSchemaProperty, domainObject: domainObjectProperty }) =>
      defineQueryFunctionInputExpressionForDomainObjectProperty({
        domainObjectName: domainObject.name,
        sqlSchemaProperty,
        domainObjectProperty,
        allSqlSchemaRelationships,
        context: GetTypescriptCodeForPropertyContext.FOR_UPSERT_QUERY,
      }),
  );

  // define the output type
  const outputType = defineOutputTypeOfFoundDomainObject(domainObject);

  // define the content
  const code = `
${imports.join('\n')}

export const sql = \`
  -- query_name = upsert_${sqlSchemaName}
  SELECT
    id, uuid
  FROM upsert_${sqlSchemaName}(
    ${queryInputExpressions.join(',\n    ')}
  );
\`;

export const upsert = async ({
  dbConnection,
  ${camelCase(domainObject.name)},
}: {
  dbConnection: DatabaseConnection;
  ${camelCase(domainObject.name)}: ${domainObject.name};
}): Promise<${outputType}> => {
  const results = await sqlQueryUpsert${domainObject.name}({
    dbExecute: dbConnection.query,
    logDebug: log.debug,
    input: {
      ${queryFunctionInputExpressions.join(',\n      ')},
    },
  });
  const { id${hasUuidProperty ? ', uuid' : ''} } = result[0]; // grab the db generated values
  return new ${domainObject.name}({ ...${camelCase(domainObject.name)}, id${
    hasUuidProperty ? ', uuid' : ''
  } }) as ${outputType};
};
`.trim();

  // return the code
  return code;
};
