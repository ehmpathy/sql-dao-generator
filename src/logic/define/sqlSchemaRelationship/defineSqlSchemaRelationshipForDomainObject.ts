import { snakeCase } from 'change-case';
import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { assertDomainObjectIsSafeToHaveSqlSchema } from './assertDomainObjectIsSafeToHaveSqlSchema';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaPropertyForDomainObjectProperty } from './defineSqlSchemaPropertyForDomainObjectProperty';
import { isNotADatabaseGeneratedProperty } from './isNotADatabaseGeneratedProperty';

export const defineSqlSchemaRelationshipForDomainObject = ({
  domainObject,
  allDomainObjects,
}: {
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaToDomainObjectRelationship => {
  // figure out some relevant info
  const sqlSchemaName = snakeCase(domainObject.name); // names are in snake case

  // define the properties
  const databaseGeneratedSqlSchemaProperties =
    defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({
      domainObject,
    });
  const propertiesRelationship = [
    /// autogenerated values
    ...databaseGeneratedSqlSchemaProperties.map((sqlSchemaProperty) => {
      // see if this property was defined on the domain object (i.e., they want the value for it)
      const domainObjectProperty =
        Object.values(domainObject.properties).find(
          (property) => snakeCase(property.name) === sqlSchemaProperty.name,
        ) ?? null;

      // if so, validate that it has the correct type
      if (domainObjectProperty) {
        const expectedType = (() => {
          if (sqlSchemaProperty.name === 'id')
            return DomainObjectPropertyType.NUMBER;
          if (sqlSchemaProperty.name === 'uuid')
            return DomainObjectPropertyType.STRING;
          if (sqlSchemaProperty.name === 'created_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'effective_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'updated_at')
            return DomainObjectPropertyType.DATE;
          throw new UnexpectedCodePathDetectedError({
            reason:
              'unexpected database generated property name. this is a bug within sql-schema-generator.', // fail fast if expectations not met
            domainObjectName: domainObject.name,
            domainObjectPropertyName: sqlSchemaProperty.name,
          });
        })();
        if (domainObjectProperty.type !== expectedType)
          throw new UserInputError({
            reason: `reserved database-generated property does not match expected type. expected ${expectedType} but received ${domainObjectProperty.type}`,
            domainObjectName: domainObject.name,
            domainObjectPropertyName: domainObjectProperty.name,
          });
      }

      // return the relationship
      return {
        sqlSchema: sqlSchemaProperty,
        domainObject: domainObjectProperty,
      };
    }),

    // non autogenerated values
    ...Object.values(domainObject.properties)
      .filter(
        (domainObjectProperty) =>
          !databaseGeneratedSqlSchemaProperties
            .map(({ name }) => name)
            .includes(snakeCase(domainObjectProperty.name)), // filter out the db generated values, because we already got them above
      )
      .map((domainObjectProperty) => {
        return {
          sqlSchema: defineSqlSchemaPropertyForDomainObjectProperty({
            property: domainObjectProperty,
            domainObject,
            allDomainObjects,
          }),
          domainObject: domainObjectProperty,
        };
      }),
  ];

  // define the unique keys
  const sqlSchemaUniqueKeys = (() => {
    // define what the unique prop names are, based on what kind of object it is
    const domainObjectPropertyNamesInUniqueKey =
      domainObject.extends === DomainObjectVariant.DOMAIN_LITERAL
        ? propertiesRelationship // if its a literal, the schema is unique on all properties of the object (excluding autogenerated ones like id and uuid)
            .filter(isNotADatabaseGeneratedProperty)
            .map(
              ({ domainObject: domainObjectProperty }) =>
                domainObjectProperty.name,
            )
        : domainObject.decorations.unique; // otherwise, its unique on what was specified

    // convert the domain object prop names into sql schema prop names
    return (
      domainObjectPropertyNamesInUniqueKey?.map((domainObjectPropertyName) => {
        const sqlSchemaPropertyName = propertiesRelationship.find(
          (propertyRelationship) =>
            propertyRelationship.domainObject?.name ===
            domainObjectPropertyName,
        )?.sqlSchema.name;
        if (!sqlSchemaPropertyName)
          throw new UserInputError({
            reason: 'Unique keys must be properties of the domain object.',
            domainObjectName: domainObject.name,
            potentialSolution: ` Was there a typo defining '${
              domainObject.name
            }.unique' with key name '${domainObjectPropertyName}'? (Is it one of ${JSON.stringify(
              Object.keys(domainObject.properties),
            )}?)`,
          });
        return sqlSchemaPropertyName;
      }) ?? null
    );
  })();

  // make sure that this domain object is safe to have a schema defined for it
  assertDomainObjectIsSafeToHaveSqlSchema({ domainObject });

  // return the relationship
  return new SqlSchemaToDomainObjectRelationship({
    name: {
      sqlSchema: sqlSchemaName,
      domainObject: domainObject.name,
    },
    properties: propertiesRelationship,
    decorations: {
      alias: {
        domainObject: domainObject.decorations.alias,
      },
      unique: {
        sqlSchema: sqlSchemaUniqueKeys,
        domainObject: domainObject.decorations.unique,
      },
    },
  });
};
