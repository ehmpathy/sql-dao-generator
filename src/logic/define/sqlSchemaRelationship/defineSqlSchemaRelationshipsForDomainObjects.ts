import { DomainObjectMetadata } from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from './defineSqlSchemaRelationshipForDomainObject';

/**
 * uses the full list of domain objects to define a relationship between its DomainObject representation and its SqlSchema representation
 */
export const defineSqlSchemaRelationshipsForDomainObjects = ({
  domainObjects,
}: {
  domainObjects: DomainObjectMetadata[];
}) =>
  domainObjects.map((domainObject) =>
    defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: domainObjects,
    }),
  );
