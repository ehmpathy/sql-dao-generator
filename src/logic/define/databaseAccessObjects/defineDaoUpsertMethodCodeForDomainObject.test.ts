import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from '../sqlSchemaRelationship/defineSqlSchemaRelationshipForDomainObject';
import { defineDaoUpsertMethodCodeForDomainObject } from './defineDaoUpsertMethodCodeForDomainObject';

describe('defineDaoUpsertMethodCodeForDomainObject', () => {
  it('should look correct for simple literal', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Geocode',
      extends: DomainObjectVariant.DOMAIN_LITERAL,
      properties: {
        latitude: {
          name: 'latitude',
          type: DomainObjectPropertyType.NUMBER,
        },
        longitude: {
          name: 'longitude',
          type: DomainObjectPropertyType.NUMBER,
        },
      },
      decorations: {
        unique: null,
        updatable: null,
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [domainObject],
    });

    // run it
    const code = defineDaoUpsertMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
      allSqlSchemaRelationships: [sqlSchemaRelationship],
    });

    // log an example
    expect(code).toContain('-- query_name = upsert_geocode'); // name of query
    expect(code).toContain(
      `
  FROM upsert_geocode(
    :latitude,
    :longitude
  )
    `.trim(),
    ); // calls upsert correctly
    expect(code).toContain(
      `
export const upsert = async (
  {
    geocode,
  }: {
    geocode: Geocode;
  },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<Geocode>>
      `.trim(),
    ); // defines fn correctly
    expect(code).toContain('await sqlQueryUpsertGeocode({');
    expect(code).toContain(
      `
    input: {
      latitude: geocode.latitude,
      longitude: geocode.longitude,
    }
      `.trim(),
    ); // defines inputs correctly
    expect(code).toMatchSnapshot();
  });
  it('should look correct for simple domain entity', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Carriage',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        cin: {
          name: 'cin',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        carries: {
          name: 'carries',
          type: DomainObjectPropertyType.ENUM,
          of: ['PASSENGER', 'FREIGHT'],
          required: true,
        },
        capacity: {
          name: 'capacity',
          type: DomainObjectPropertyType.NUMBER,
          nullable: true,
        },
      },
      decorations: {
        unique: ['cin'],
        updatable: ['capacity'],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [domainObject],
    });

    // run it
    const code = defineDaoUpsertMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
      allSqlSchemaRelationships: [sqlSchemaRelationship],
    });

    // log an example
    expect(code).toContain('-- query_name = upsert_carriage'); // name of query
    expect(code).toContain(
      `
  FROM upsert_carriage(
    :cin,
    :carries,
    :capacity
  )
    `.trim(),
    ); // calls upsert correctly
    expect(code).toContain(
      `
async (
  {
    carriage,
  }: {
    carriage: Carriage;
  },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<Carriage>>
      `.trim(),
    ); // defines fn correctly
    expect(code).toContain('await sqlQueryUpsertCarriage({');
    expect(code).toContain(
      `
    input: {
      cin: carriage.cin,
      carries: carriage.carries,
      capacity: carriage.capacity,
    }
      `.trim(),
    ); // defines inputs correctly
    expect(code).toMatchSnapshot();
  });
  it('should return the exact autogenerated values that were defined on the domain entity', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Carriage',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        effectiveAt: {
          name: 'effectiveAt',
          type: DomainObjectPropertyType.DATE,
          required: false,
        },
        cin: {
          name: 'cin',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        carries: {
          name: 'carries',
          type: DomainObjectPropertyType.ENUM,
          of: ['PASSENGER', 'FREIGHT'],
          required: true,
        },
        capacity: {
          name: 'capacity',
          type: DomainObjectPropertyType.NUMBER,
          nullable: true,
        },
      },
      decorations: {
        unique: ['cin'],
        updatable: ['capacity'],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [domainObject],
    });

    // run it
    const code = defineDaoUpsertMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
      allSqlSchemaRelationships: [sqlSchemaRelationship],
    });

    // log an example
    expect(code).toContain('-- query_name = upsert_carriage'); // name of query
    expect(code).toContain('dgv.id, dgv.uuid, dgv.effective_at'); // name of query
    expect(code).toContain(
      `
  FROM upsert_carriage(
    :cin,
    :carries,
    :capacity
  )
    `.trim(),
    ); // calls upsert correctly
    expect(code).toContain(
      `
async (
  {
    carriage,
  }: {
    carriage: Carriage;
  },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<Carriage>>
      `.trim(),
    ); // defines fn correctly
    expect(code).toContain('await sqlQueryUpsertCarriage({');
    expect(code).toContain(
      `
    input: {
      cin: carriage.cin,
      carries: carriage.carries,
      capacity: carriage.capacity,
    }
      `.trim(),
    ); // defines inputs correctly
    expect(code).toContain('id, uuid, effective_at: effectiveAt');
    expect(code).toContain(
      'return new Carriage({ ...carriage, id, uuid, effectiveAt })',
    );
    expect(code).toMatchSnapshot();
  });
  it('should look correct for a domain event with a static referenced array', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'TrainLocatedEvent',
      extends: DomainObjectVariant.DOMAIN_EVENT,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        trainUuid: {
          name: 'trainUuid',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        occurredAt: {
          name: 'occurredAt',
          type: DomainObjectPropertyType.DATE,
          required: true,
        },
        geocodes: {
          name: 'geocodes',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
          required: true,
        },
      },
      decorations: {
        unique: ['trainUuid', 'occurredAt'],
        updatable: [],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [] as DomainObjectMetadata[],
    });

    // define property that gets referenced
    const geocodeSqlSchemaRelationship =
      defineSqlSchemaRelationshipForDomainObject({
        domainObject: new DomainObjectMetadata({
          name: 'Geocode',
          extends: DomainObjectVariant.DOMAIN_LITERAL,
          properties: {
            latitude: {
              name: 'latitude',
              type: DomainObjectPropertyType.NUMBER,
            },
            longitude: {
              name: 'longitude',
              type: DomainObjectPropertyType.NUMBER,
            },
          },
          decorations: {
            unique: null,
            updatable: null,
          },
        }),
        allDomainObjects: [domainObject],
      });

    // run it
    const code = defineDaoUpsertMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
      allSqlSchemaRelationships: [
        sqlSchemaRelationship,
        geocodeSqlSchemaRelationship,
      ],
    });

    // log an example
    expect(code).toContain('-- query_name = upsert_train_located_event'); // name of query
    expect(code).toContain(
      `
  FROM upsert_train_located_event(
    :trainUuid,
    :occurredAt,
    :geocodeIds
  )
    `.trim(),
    ); // calls upsert correctly
    expect(code).toContain(
      `
async (
  {
    trainLocatedEvent,
  }: {
    trainLocatedEvent: TrainLocatedEvent;
  },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<TrainLocatedEvent>>
      `.trim(),
    ); // defines fn correctly
    expect(code).toContain('await sqlQueryUpsertTrainLocatedEvent({');
    expect(code).toContain(
      `
    input: {
      trainUuid: trainLocatedEvent.trainUuid,
      occurredAt: trainLocatedEvent.occurredAt,
      geocodeIds: await Promise.all(trainLocatedEvent.geocodes.map(async (geocode) => geocode.id ? geocode.id : (await geocodeDao.upsert({ geocode }, context)).id)),
    }
      `.trim(),
    ); // defines inputs correctly
    expect(code).toMatchSnapshot();
  });
  it('should look correct for domain entity with references, array and solo, implicit and direct', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Train',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        tin: {
          name: 'tin',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        homeStationGeocode: {
          name: 'homeStationGeocode',
          type: DomainObjectPropertyType.REFERENCE,
          of: {
            name: 'Geocode',
            extends: DomainObjectVariant.DOMAIN_LITERAL,
          },
        },
        leadEngineerUuid: {
          name: 'leadEngineerUuid',
          type: DomainObjectPropertyType.STRING,
        },
        badges: {
          name: 'badges',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'TrainBadge',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        locomotiveUuids: {
          name: 'locomotiveUuids',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.STRING,
          },
        },
      },
      decorations: {
        unique: ['tin'],
        updatable: [
          'homeStation',
          'badges',
          'locomotiveUuids',
          'leadEngineerUuid',
        ],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [
        domainObject,
        { name: 'Locomotive', extends: DomainObjectVariant.DOMAIN_ENTITY },
        { name: 'TrainEngineer', extends: DomainObjectVariant.DOMAIN_ENTITY },
      ] as DomainObjectMetadata[],
    });

    // define property that gets referenced
    const geocodeSqlSchemaRelationship =
      defineSqlSchemaRelationshipForDomainObject({
        domainObject: new DomainObjectMetadata({
          name: 'Geocode',
          extends: DomainObjectVariant.DOMAIN_LITERAL,
          properties: {
            latitude: {
              name: 'latitude',
              type: DomainObjectPropertyType.NUMBER,
            },
            longitude: {
              name: 'longitude',
              type: DomainObjectPropertyType.NUMBER,
            },
          },
          decorations: {
            unique: null,
            updatable: null,
          },
        }),
        allDomainObjects: [domainObject],
      });
    const badgeSqlSchemaRelationship =
      defineSqlSchemaRelationshipForDomainObject({
        domainObject: new DomainObjectMetadata({
          name: 'TrainBadge',
          extends: DomainObjectVariant.DOMAIN_LITERAL,
          properties: {
            name: {
              name: 'name',
              type: DomainObjectPropertyType.STRING,
            },
            description: {
              name: 'description',
              type: DomainObjectPropertyType.STRING,
            },
            rank: {
              name: 'rank',
              type: DomainObjectPropertyType.ENUM,
              of: ['GOLD', 'SILVER', 'BRONZE'],
            },
          },
          decorations: {
            unique: null,
            updatable: null,
          },
        }),
        allDomainObjects: [domainObject],
      });
    const locomotiveSqlSchemaRelationship =
      defineSqlSchemaRelationshipForDomainObject({
        domainObject: new DomainObjectMetadata({
          name: 'Locomotive',
          extends: DomainObjectVariant.DOMAIN_ENTITY,
          properties: {
            uuid: { name: 'uuid', type: DomainObjectPropertyType.STRING },
          },
          decorations: { unique: ['uuid'], updatable: [] },
        }),
        allDomainObjects: [domainObject],
      });
    const engineerSqlSchemaRelationship =
      defineSqlSchemaRelationshipForDomainObject({
        domainObject: new DomainObjectMetadata({
          name: 'TrainEngineer',
          extends: DomainObjectVariant.DOMAIN_ENTITY,
          properties: {
            uuid: { name: 'uuid', type: DomainObjectPropertyType.STRING },
          }, // domain entity reference, so we dont need to look at properties
          decorations: { unique: ['uuid'], updatable: [] },
        }),
        allDomainObjects: [domainObject],
      });

    // run it
    const code = defineDaoUpsertMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
      allSqlSchemaRelationships: [
        sqlSchemaRelationship,
        geocodeSqlSchemaRelationship,
        badgeSqlSchemaRelationship,
        locomotiveSqlSchemaRelationship,
        engineerSqlSchemaRelationship,
      ],
    });

    // log an example
    expect(code).toContain('-- query_name = upsert_train'); // name of query
    expect(code).toContain(
      `
  FROM upsert_train(
    :tin,
    :homeStationGeocodeId,
    (SELECT id FROM train_engineer WHERE train_engineer.uuid = :leadEngineerUuid),
    :badgeIds,
    (
      SELECT COALESCE(array_agg(locomotive.id ORDER BY locomotive_ref.array_order_index), array[]::bigint[]) AS array_agg
      FROM locomotive
      JOIN unnest(:locomotiveUuids::uuid[]) WITH ORDINALITY
        AS locomotive_ref (uuid, array_order_index)
        ON locomotive.uuid = locomotive_ref.uuid
    )
  )
    `.trim(),
    ); // calls upsert correctly
    expect(code).toContain(
      `
async (
  {
    train,
  }: {
    train: Train;
  },
  context: { dbConnection: DatabaseConnection },
): Promise<HasMetadata<Train>>
      `.trim(),
    ); // defines fn correctly
    expect(code).toContain('await sqlQueryUpsertTrain({');
    expect(code).toContain(
      `
    input: {
      tin: train.tin,
      homeStationGeocodeId: train.homeStationGeocode.id ? train.homeStationGeocode.id : (await geocodeDao.upsert({ geocode: train.homeStationGeocode }, context)).id,
      leadEngineerUuid: train.leadEngineerUuid,
      badgeIds: await Promise.all(train.badges.map(async (trainBadge) => trainBadge.id ? trainBadge.id : (await trainBadgeDao.upsert({ trainBadge }, context)).id)),
      locomotiveUuids: train.locomotiveUuids,
    }
      `.trim(),
    ); // defines inputs correctly
    expect(code).toMatchSnapshot();
  });
});
