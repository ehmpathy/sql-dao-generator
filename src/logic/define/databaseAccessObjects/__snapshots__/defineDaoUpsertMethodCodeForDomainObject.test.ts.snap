// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineDaoUpsertMethodCodeForDomainObject should look correct for a domain event with a static referenced array 1`] = `
"import { HasId } from 'simple-type-guards';

import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';
import { TrainLocatedEvent } from '$PATH_TO_DOMAIN_OBJECT';
import { log } from '$PATH_TO_LOG_OBJECT';
import { sqlQueryUpsertTrainLocatedEvent } from '$PATH_TO_GENERATED_SQL_QUERY_FUNCTIONS';
import { geocodeDao } from '../geocodeDao';

export const sql = \`
  -- query_name = upsert_train_located_event
  SELECT
    dgv.id, dgv.uuid
  FROM upsert_train_located_event(
    :trainUuid,
    :occurredAt,
    :geocodeIds
  ) as dgv;
\`;

export const upsert = async ({
  dbConnection,
  trainLocatedEvent,
}: {
  dbConnection: DatabaseConnection;
  trainLocatedEvent: TrainLocatedEvent;
}): Promise<HasId<TrainLocatedEvent>> => {
  const results = await sqlQueryUpsertTrainLocatedEvent({
    dbExecute: dbConnection.query,
    logDebug: log.debug,
    input: {
      trainUuid: trainLocatedEvent.trainUuid,
      occurredAt: trainLocatedEvent.occurredAt,
      geocodeIds: await Promise.all(trainLocatedEvent.geocodes.map(async (geocode) => geocode.id ? geocode.id : (await geocodeDao.upsert({ dbConnection, geocode })).id)),
    },
  });
  const { id } = results[0]; // grab the db generated values
  return new TrainLocatedEvent({ ...trainLocatedEvent, id }) as HasId<TrainLocatedEvent>;
};"
`;

exports[`defineDaoUpsertMethodCodeForDomainObject should look correct for domain entity with references, array and solo, implicit and direct 1`] = `
"import { HasId, HasUuid } from 'simple-type-guards';

import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';
import { Train } from '$PATH_TO_DOMAIN_OBJECT';
import { log } from '$PATH_TO_LOG_OBJECT';
import { sqlQueryUpsertTrain } from '$PATH_TO_GENERATED_SQL_QUERY_FUNCTIONS';
import { geocodeDao } from '../geocodeDao';
import { trainBadgeDao } from '../trainBadgeDao';

export const sql = \`
  -- query_name = upsert_train
  SELECT
    dgv.id, dgv.uuid
  FROM upsert_train(
    :tin,
    :homeStationGeocodeId,
    (SELECT id FROM train_engineer WHERE train_engineer.uuid = :leadEngineerUuid),
    :badgeIds,
    (
      SELECT COALESCE(array_agg(locomotive.id ORDER BY locomotive_ref.array_order_index), array[]::bigint[]) AS array_agg
      FROM locomotive
      JOIN unnest(:locomotiveUuids) WITH ORDINALITY
        AS locomotive_ref (uuid, array_order_index)
        ON locomotive.uuid = locomotive_ref.uuid
    )
  ) as dgv;
\`;

export const upsert = async ({
  dbConnection,
  train,
}: {
  dbConnection: DatabaseConnection;
  train: Train;
}): Promise<HasId<HasUuid<Train>>> => {
  const results = await sqlQueryUpsertTrain({
    dbExecute: dbConnection.query,
    logDebug: log.debug,
    input: {
      tin: train.tin,
      homeStationGeocodeId: train.homeStationGeocode.id ? train.homeStationGeocode.id : (await geocodeDao.upsert({ dbConnection, geocode: train.homeStationGeocode })).id,
      leadEngineerUuid: train.leadEngineerUuid,
      badgeIds: await Promise.all(train.badges.map(async (trainBadge) => trainBadge.id ? trainBadge.id : (await trainBadgeDao.upsert({ dbConnection, trainBadge })).id)),
      locomotiveUuids: train.locomotiveUuids,
    },
  });
  const { id, uuid } = results[0]; // grab the db generated values
  return new Train({ ...train, id, uuid }) as HasId<HasUuid<Train>>;
};"
`;

exports[`defineDaoUpsertMethodCodeForDomainObject should look correct for simple domain entity 1`] = `
"import { HasId, HasUuid } from 'simple-type-guards';

import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';
import { Carriage } from '$PATH_TO_DOMAIN_OBJECT';
import { log } from '$PATH_TO_LOG_OBJECT';
import { sqlQueryUpsertCarriage } from '$PATH_TO_GENERATED_SQL_QUERY_FUNCTIONS';

export const sql = \`
  -- query_name = upsert_carriage
  SELECT
    dgv.id, dgv.uuid
  FROM upsert_carriage(
    :cin,
    :carries,
    :capacity
  ) as dgv;
\`;

export const upsert = async ({
  dbConnection,
  carriage,
}: {
  dbConnection: DatabaseConnection;
  carriage: Carriage;
}): Promise<HasId<HasUuid<Carriage>>> => {
  const results = await sqlQueryUpsertCarriage({
    dbExecute: dbConnection.query,
    logDebug: log.debug,
    input: {
      cin: carriage.cin,
      carries: carriage.carries,
      capacity: carriage.capacity,
    },
  });
  const { id, uuid } = results[0]; // grab the db generated values
  return new Carriage({ ...carriage, id, uuid }) as HasId<HasUuid<Carriage>>;
};"
`;

exports[`defineDaoUpsertMethodCodeForDomainObject should look correct for simple value object 1`] = `
"import { HasId } from 'simple-type-guards';

import { DatabaseConnection } from '$PATH_TO_DATABASE_CONNECTION';
import { Geocode } from '$PATH_TO_DOMAIN_OBJECT';
import { log } from '$PATH_TO_LOG_OBJECT';
import { sqlQueryUpsertGeocode } from '$PATH_TO_GENERATED_SQL_QUERY_FUNCTIONS';

export const sql = \`
  -- query_name = upsert_geocode
  SELECT
    dgv.id, dgv.uuid
  FROM upsert_geocode(
    :latitude,
    :longitude
  ) as dgv;
\`;

export const upsert = async ({
  dbConnection,
  geocode,
}: {
  dbConnection: DatabaseConnection;
  geocode: Geocode;
}): Promise<HasId<Geocode>> => {
  const results = await sqlQueryUpsertGeocode({
    dbExecute: dbConnection.query,
    logDebug: log.debug,
    input: {
      latitude: geocode.latitude,
      longitude: geocode.longitude,
    },
  });
  const { id } = results[0]; // grab the db generated values
  return new Geocode({ ...geocode, id }) as HasId<Geocode>;
};"
`;
